### ISSUE TITLE
Build benchmarking infrastructure to compare parallel speed ups

### ISSUE THREAD

**Issue by MridulS** [networkx/nx-parallel#12]
We need to have a quick way of using either github actions or scripts to run some crude benchmarks while developing new algorithms.

**Comment by MridulS** [comment]
The current benchmarking scripts are not too portable, they should run in some automated fashion.

**Comment by Schefflera-Arboricola** [comment]
@MridulS I'd like to take up this issue, but I don't have any experience in building benchmarking infrastructures, so please guide me on this.

### A possible benchmarking approach : 
we can check if the average of all the speedup values(for graphs of different sizes and densities) is greater than 1, to ensure that parallel algos are more time efficient or we can also use `pytest-benchmark` like below.

`test_benchmarks.py`
```.py
import pytest
import networkx as nx
import nx_parallel

num,p=300,0.5
G = nx.fast_gnp_random_graph(num, p, directed=False)
H = nx_parallel.ParallelGraph(G)

@pytest.mark.benchmark
def test_algorithm_performance_G(benchmark): 
    result_seq = benchmark(nx.betweenness_centrality, G) #replace "betweenness_centrality" with the new function added

@pytest.mark.benchmark
def test_algorithm_performance_H(benchmark):
    result_para = benchmark(nx.betweenness_centrality, H)
```

`.github/workflows/benchmark.yml` : 
```.yml
name: Benchmark

on:
  push:
    branches:
      - main 

jobs:
  benchmark:
    runs-on: ubuntu-latest

    steps:
    - name: Checkout code
      uses: actions/checkout@v2

    - name: Set up Python
      uses: actions/setup-python@v2
      with:
        python-version: 3.9

    - name: Install dependencies
      run: pip install networkx pytest pytest-benchmark nx_parallel

    - name: Run benchmarks
      run: pytest test_benchmark.py 
```

Test output : 

<img width="1440" alt="Screenshot 2023-10-14 at 7 04 28 PM" src="https://github.com/networkx/nx-parallel/assets/91629733/a45b57e4-e7c2-436e-86dd-1df87fa3a013">


What are your thoughts on this? What all should I keep in mind before structuring it?

Thank you :)

**Comment by MridulS** [comment]
@Schefflera-Arboricola yes, pytest-benchmark could be one way of doing this. We use [ASV](https://asv.readthedocs.io/en/v0.6.1/) for networkx benchmarking. But it's possible we would need to come up with a way which incorporates benchmarks with networkx dispatching. Ideally this benchmark suite will be able to swap in any backend (graphblas, nx-parallel, cugraph) and run it against all of them. We still need to think a bit more about how to approach this :)

**Comment by Schefflera-Arboricola** [comment]
> We need to have a quick way of using either github actions....

by "github actions" did you mean something like this: https://github.com/networkx/networkx/pull/6834 ? or something else?

**Comment by MridulS** [comment]
Yes! I'll try to finish the one in NX main repo soon. I think it's already good to go.

**Comment by Schefflera-Arboricola** [comment]
Just adding this for reference here: https://conbench.github.io/conbench/
